src/stan/mcmc/hmc/nuts_classic/base_nuts_classic.hpp

```diff
@@ -0,0 +1,249 @@
+                                           writer);
+          ++(this->depth_);
+
+          *z = this->z_;
+
+          // Metropolis-Hastings sample the fresh subtree
+          if (!util.criterion)
+            break;
+
+          double subtree_prob = 0;
+
+          if (n_valid) {
+            subtree_prob = static_cast<double>(n_valid_subtree) /
+              static_cast<double>(n_valid);
+          } else {
+            subtree_prob = n_valid_subtree ? 1 : 0;
+          }
+
+          if (this->rand_uniform_() < subtree_prob)
+            z_sample = z_propose;
+
+          n_valid += n_valid_subtree;
+
+          // Check validity of completed tree
+          this->z_.ps_point::operator=(z_plus);
+          Eigen::VectorXd delta_rho = rho_minus + rho_init + rho_plus;
+
+          util.criterion = compute_criterion(z_minus, this->z_, delta_rho);
+        }
+
+        this->n_leapfrog_ = util.n_tree;
+
+        double accept_prob = util.sum_prob / static_cast<double>(util.n_tree);
+
+        this->z_.ps_point::operator=(z_sample);
+        this->energy_ = this->hamiltonian_.H(this->z_);
+        return sample(this->z_.q, - this->z_.V, accept_prob);
+      }
+
+      void get_sampler_param_names(std::vector<std::string>& names) {
+        names.push_back("stepsize__");
+        names.push_back("treedepth__");
+        names.push_back("n_leapfrog__");
+        names.push_back("divergent__");
+        names.push_back("energy__");
+      }
+
+      void get_sampler_params(std::vector<double>& values) {
+        values.push_back(this->epsilon_);
+        values.push_back(this->depth_);
+        values.push_back(this->n_leapfrog_);
+        values.push_back(this->divergent_);
+        values.push_back(this->energy_);
+      }
+
+      virtual bool compute_criterion(ps_point& start,
+                                     typename Hamiltonian<Model, BaseRNG>
+                                     ::PointType& finish,
+                                     Eigen::VectorXd& rho) = 0;
+
+      // Returns number of valid points in the completed subtree
+      int build_tree(int depth, Eigen::VectorXd& rho,
+                     ps_point* z_init_parent, ps_point& z_propose,
+                     nuts_util& util,
+                     interface_callbacks::writer::base_writer& writer) {
+        // Base case
+        if (depth == 0) {
+            this->integrator_.evolve(this->z_, this->hamiltonian_,
+                                     util.sign * this->epsilon_,
+                                     writer);
+            rho += this->z_.p;
+
+            if (z_init_parent) *z_init_parent = this->z_;
+            z_propose = this->z_;
+
+            double h = this->hamiltonian_.H(this->z_);
+            if (boost::math::isnan(h))
+              h = std::numeric_limits<double>::infinity();
+
+            util.criterion = util.log_u + (h - util.H0) < this->max_delta_;
+            if (!util.criterion) ++(this->divergent_);
```

Comment:
Shouldn't `divergent_` be `bool` rather than `int` now?  In which case, the line above can just set the value to `true`.